### PR TITLE
Disarm refactor, plus shoving people with shields

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
I wanted to add the ability to shove people with shields by right-clicking your target, just like how it works barehanded.

This also required a solid refactor of disarm code, effectively bringing down the core of it to `mob/living` from `mob/living/carbon` or `mob/living/carbon/human`. This also means you can shove simple mobs inside closets, bins and on tables.

Xenos and borgs are pretty much immune to regular disarms, but using a shield will work (borgs and royal xenos are immune to the knockdown).

The riot shield armor has been balanced. It now tanks melee attacks pretty well, but will break against bullets in just about 2 to 4 hits depending on the bullet damage. I've always found the lack of sturdiness of the riot shields for what they're supposed to be good for a bit detrimental.

Because I've refactored an item flag into a trait, I've had to add a new MOD module that grants protection from shove knockdown and staggering; found pre-installed in the administrative MODsuit, but I've also added it to the black market to make it cooler.

You can bash people with the strobe shield on combat mode.

## Why It's Good For The Game
Currently, shields are simply items that take a held slot in return of some block chance without being anything special, save for the strobe shield's integrated flash I guess, but are also a botherance as most crumple under the duress of less than half a dozen attacks. Meanwhile swords and other weapons with blok chance just don't care.
TL;DR, I want them a bit more remarkable, and flexible as a tool.

Of course, this ended up in a larger refactor because the right-click / disarm code was inconsistent.

## Changelog

:cl:
add: Shields (and pillows) can be used to shove people around the same way barehanded right-clicking does. Xenos and borgs can actually be moved this way.
add: Added a new MODsuit module, the bulwark module, which prevents knockdown and staggering from shoving, and getting pushed away by thrown objects. Inbuilt for the safeguard MODsuit, but one might also it in the black market.
refactor: Disarming has been refactored. You can now shove simple critters onto tables and into bins and closets
balance: Shields now take their own armor values and the armor penetration of the attack they blocked when damaged. This means shields are a bit sturdier now.
balance: Riot shields can tank a lot more damage against melee weapons, but less against bullets.
qol: strobe shields can now be used to bash people while combat mode is on.
/:cl:

### CODE REVIEW

**Comment by Loafin34** [comment]
maybe add this to security modsuits so they aren't downgrades of existing armor?

**Comment by DrDiasyl** [comment]
Shoving with shield should make some meatier sound and maybe even deal some damage, at least I think this would be cool and add more Police brutality to the Security (as they love it that much)

**Comment by DrDiasyl** [comment]
Also, does it also now tank against thrown items? It counts as melee, right?
Riot shields should be very good against thrown items.

**Comment by Dawnseer** [comment]
We're getting closer to getting into the testudo formation with the homies

File: code/game/objects/items/shields.dm [hunk 1]
```diff
@@ -339,4 +360,11 @@
 	playsound(src, 'sound/weapons/batonextend.ogg', 50, TRUE)
 	return COMPONENT_NO_DEFAULT_MESSAGE
 
+/obj/item/shield/riot/proc/can_disarm_attack(datum/source, mob/living/victim, mob/living/user, send_message)
```
Comment (necromanceranne) [inline code/game/objects/items/shields.dm:363]:
```suggestion
/obj/item/shield/riot/tele/proc/can_disarm_attack(datum/source, mob/living/victim, mob/living/user, send_message)
```

File: code/modules/clothing/suits/armor.dm [hunk 2]
```diff
@@ -253,10 +253,18 @@
...
 	heat_protection = CHEST|GROIN|LEGS|FEET|ARMS|HANDS
 	armor_type = /datum/armor/armor_riot
-	clothing_flags = BLOCKS_SHOVE_KNOCKDOWN
 	strip_delay = 80
 	equip_delay_other = 60
 
+/obj/item/clothing/suit/armor/riot/equipped(mob/living/user, slot)
+	. = ..()
+	if(slot & ITEM_SLOT_OCLOTHING)
+		ADD_TRAIT(user, TRAIT_SHOVE_KNOCKDOWN_BLOCKED, REF(src))
```
Comment (MrMelbert) [inline code/modules/clothing/suits/armor.dm:262]: use `clothing_traits`

File: code/modules/clothing/suits/armor.dm [hunk 3]
```diff
@@ -373,6 +381,15 @@
 	acid = 100
 	wound = 15
 
+/obj/item/clothing/suit/armor/swat/equipped(mob/living/user, slot)
+	. = ..()
+	if(slot & ITEM_SLOT_OCLOTHING)
+		ADD_TRAIT(user, TRAIT_SHOVE_KNOCKDOWN_BLOCKED, REF(src))
```
Comment (MrMelbert) [inline code/modules/clothing/suits/armor.dm:387]: use `clothing_traits`

File: code/datums/elements/disarm_attack.dm [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	return COMPONENT_SECONDARY_CANCEL_ATTACK_CHAIN
+
+///check if the item conditions for the disarm action are met.
+/datum/element/disarm_attack/proc/can_disarm_attack(obj/item/source, mob/living/victim, mob/living/user, message = TRUE)
+	if(SEND_SIGNAL(source, COMSIG_ITEM_CAN_DISARM_ATTACK, victim, user) & COMPONENT_BLOCK_ITEM_DISARM_ATTACK)
+		return FALSE
+	return TRUE
+
+/datum/element/disarm_attack/proc/examine(obj/item/source, mob/user, list/examine_list)
+	if(can_disarm_attack(source, null, user, FALSE))
```
Comment (MrMelbert) [inline code/datums/elements/disarm_attack.dm:29]:
```suggestion
/datum/element/disarm_attack/proc/examine(obj/item/source, mob/user, list/examine_list)
	SIGNAL_HANDLER
	if(can_disarm_attack(source, null, user, FALSE))
```

File: code/game/objects/items/shields.dm [hunk 5]
```diff
@@ -57,21 +61,17 @@
...
 		return TRUE
-	if (atom_integrity <= damage)
-		var/turf/owner_turf = get_turf(owner)
-		owner_turf.visible_message(span_warning("[hitby] destroys [src]!"))
-		shatter(owner)
-		qdel(src)
-		return FALSE
-	take_damage(damage)
-	return TRUE
+	take_damage(damage, damage_type, attack_type)
```
Comment (necromanceranne) [inline code/game/objects/items/shields.dm:67]: Unfortunately, ``attack_type`` doesn't aligns with armor flag categories. I think you'll need to determine this separately. Since only projectiles have varying armor targets, we can just check the armor flag of the projectile if it is a ranged attack, and for anything else we can conveniently assume ``MELEE`` to be a workable default.

Comment (Ghommie) [reply code/game/objects/items/shields.dm]: Done that. I've also added armor penetration into account, at the cost of a few `else if` calls.

**Comment by Ghommie** [comment]
> Shoving with shield should make some meatier sound and maybe even deal some damage...

I did try using the shield block sound, but I couldn't decide if it was good enough. As for the damage, I'd prefer for it to stay a non-lethal option. Left-clicking does it already.

> Also, does it also now tank against thrown items? It counts as melee, right? 

Yep.

> (And also, how about making it block being thrown back after getting hit with a heavy thrown item?)

I will get to it.

> Also, will this work with pillow shields?

Done. Thanks for the suggestion.

> maybe add this to security modsuits so they aren't downgrades of existing armor?

You mean the bulwark module? Perhaps. I've been thinking to add it to the HoS' modsuit; maybe I should.

**Comment by necromanceranne** [comment]
Maybe just drop the complexity of the module itself? It's rare onto itself.

File: code/modules/mod/modules/modules_security.dm [hunk 6]
```diff
@@ -578,3 +578,20 @@
...
+	complexity = 3
+	incompatible_modules = list(/obj/item/mod/module/shove_blocker)
+
+/obj/item/mod/module/shove_blocker/on_suit_activation()
+	mod.wearer.add_traits(list(TRAIT_SHOVE_KNOCKDOWN_BLOCKED, TRAIT_SHOVE_NO_STAGGER, TRAIT_NO_THROW_HITPUSH), MOD_TRAIT)
+
+/obj/item/mod/module/shove_blocker/on_suit_deactivation(deleting = FALSE)
+	mod.wearer.remove_traits(list(TRAIT_SHOVE_KNOCKDOWN_BLOCKED, TRAIT_SHOVE_NO_STAGGER, TRAIT_NO_THROW_HITPUSH), MOD_TRAIT)
+
+/obj/item/mod/module/shove_blocker/locked
```
Comment (Fikou) [inline code/modules/mod/modules/modules_security.dm:595]: different name so people can tell its inbuilt would be cool

**Comment by Ghommie** [comment]
It seems like no one is bothering to review this. Oh, the holidays...

File: code/datums/elements/disarm_attack.dm [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+///check if the item conditions for the disarm action are met.
+/datum/element/disarm_attack/proc/can_disarm_attack(obj/item/source, mob/living/victim, mob/living/user, message = TRUE)
+	if(SEND_SIGNAL(source, COMSIG_ITEM_CAN_DISARM_ATTACK, victim, user) & COMPONENT_BLOCK_ITEM_DISARM_ATTACK)
+		return FALSE
+	return TRUE
+
+/datum/element/disarm_attack/proc/examine(obj/item/source, mob/user, list/examine_list)
+	SIGNAL_HANDLER
+	if(can_disarm_attack(source, null, user, FALSE))
+		examine_list += span_notice("You can use it to <b>shove</b> people with <b>right-click</b>.")
```
Comment (MrMelbert) [inline code/datums/elements/disarm_attack.dm:35]: I find this kinda sus, particularly that this sends the signal with a `null` victim.
I guarantee most signal consumers will not null check it.
Maybe pass user as victim, essentially checking that you can (in theory) shove yourself with it (ignoring the fact that you cannot shove yourself)?

File: code/datums/elements/disarm_attack.dm [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+		return COMPONENT_SECONDARY_CANCEL_ATTACK_CHAIN
+	if(victim.check_block(source, 0, "the [source.name]", MELEE_ATTACK, 0))
+		return COMPONENT_SECONDARY_CANCEL_ATTACK_CHAIN
+	user.disarm(victim, source)
+	user.changeNext_move(source.secondary_attack_speed || source.attack_speed)
+	return COMPONENT_SECONDARY_CANCEL_ATTACK_CHAIN
+
+///check if the item conditions for the disarm action are met.
+/datum/element/disarm_attack/proc/can_disarm_attack(obj/item/source, mob/living/victim, mob/living/user, message = TRUE)
+	if(SEND_SIGNAL(source, COMSIG_ITEM_CAN_DISARM_ATTACK, victim, user) & COMPONENT_BLOCK_ITEM_DISARM_ATTACK)
```
Comment (MrMelbert) [inline code/datums/elements/disarm_attack.dm:28]:
```suggestion
	if(SEND_SIGNAL(source, COMSIG_ITEM_CAN_DISARM_ATTACK, victim, user, message) & COMPONENT_BLOCK_ITEM_DISARM_ATTACK)
```
The docs implied this arg should be sent but you did not send it.

**Review by MrMelbert — APPROVED** [review]
Looks fine, feels like some of the disarm flags are superfluous but better to have and not need than need and not have. Small things.

File: code/datums/elements/disarm_attack.dm (hunk 7 above)
Comment (MrMelbert) [reply code/datums/elements/disarm_attack.dm:35]: Also just thinking about it the signal is kinda sus in general, maybe this should be a component with a callback `can_disarm`?

Comment (Ghommie) [reply code/datums/elements/disarm_attack.dm:35]: I think the signal is ok, but you're kinda right the null victim seems a bit janky. I followed your suggestion of passing the `user` as `victim` so we don't have to worry about petty null issues.
